stdlib/public/core/StringGutsRangeReplaceable.swift

```diff
@@ -155,22 +158,51 @@ extension _StringGuts {
 
     append(_StringGutsSlice(other))
   }
+  
+  @inline(never)
+  @_effects(readonly)
+  private func _foreignConvertedToSmall() -> _SmallString {
+    let smol = String(uninitializedCapacity: _SmallString.capacity) { buffer in
+      guard let count = _foreignCopyUTF8(into: buffer) else {
+        fatalError("String capacity was smaller than required")
+      }
+      return count
+    }
+    _internalInvariant(smol._guts.isSmall)
+    return smol._guts.asSmall
+  }
+  
+  private func _convertedToSmall() -> _SmallString {
+    _internalInvariant(utf8Count <= _SmallString.capacity)
+    if _fastPath(isSmall) {
+      return asSmall
+    }
+    if isFastUTF8 {
+      return withFastUTF8 { _SmallString($0)! }
+    }
+    return _foreignConvertedToSmall()
+  }
 
   internal mutating func append(_ slicedOther: _StringGutsSlice) {
     defer { self._invariantCheck() }
 
-    if self.isSmall && slicedOther._guts.isSmall {
+    let otherCount = slicedOther.utf8Count
+    let totalCount = utf8Count + otherCount
+    let hasEnoughUsableSpace = isUniqueNative &&
+      nativeUnusedCapacity! >= otherCount
+    let shouldBeSmol = totalCount <= _SmallString.capacity &&
+      (isSmall || !hasEnoughUsableSpace)
```

Comment:
Could you explain the logic here, ideally in a comment. There's a fair amount of and, ors, and nots.